### PR TITLE
[Merged by Bors] - feat: the derived category of an abelian category

### PR DESCRIPTION
**Pull request by joelriou**
The derived category of an abelian category is defined and it is shown that it is a triangulated category.

---
- [x] depends on: #11740
- [x] depends on: #11789
- [x] depends on: #11759
- [x] depends on: #11805
- [x] depends on: #11764
- [x] depends on: #11782
- [x] depends on: #11760
- [x] depends on: #11721
- [x] depends on: #11737
- [x] depends on: #11728
- [x] depends on: #11738
- [x] depends on: #11786 
- [x] depends on: #13558

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Triangulated/HomologicalFunctor.lean [hunk 1]
```diff
@@ -230,7 +230,7 @@ lemma mem_homologicalKernel_W_iff {X Y : C} (f : X ⟶ Y) :
   simp only [mem_homologicalKernel_iff, h₁, ← h₂, ← h₃]
   constructor
   · intro h n
-    obtain ⟨m, rfl⟩ : ∃ (m : ℤ), n = m + 1 := ⟨n - 1, by omega⟩
+    obtain ⟨m, rfl⟩ : ∃ (m : ℤ), n = m + 1 := ⟨n - 1, by simp⟩
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Triangulated/HomologicalFunctor.lean:233]: (Note: with `omega`, irrelevant unnecessary variables were added to the lemma. Using `simp` solves this issue.)

File: Mathlib/Algebra/Homology/DerivedCategory/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+def Q : CochainComplex C ℤ ⥤ DerivedCategory C := HomologicalComplexUpToQuasiIso.Q
+
+instance : (Q (C := C)).IsLocalization
+    (HomologicalComplex.quasiIso C (ComplexShape.up ℤ)) := by
+  dsimp only [Q, DerivedCategory]
+  infer_instance
+
+/-- The localization functor `HomotopyCategory C (ComplexShape.up ℤ) ⥤ DerivedCategory C`. -/
+def Qh : HomotopyCategory C (ComplexShape.up ℤ) ⥤ DerivedCategory C :=
+  HomologicalComplexUpToQuasiIso.Qh
```
Comment (erdOne) [inline Mathlib/Algebra/Homology/DerivedCategory/Basic.lean:137]: Do we want to make `C` explicit in `Q` and `Qh`? Seems that there are lots of `Q (C := C)` below.

Comment (joelriou) [reply Mathlib/Algebra/Homology/DerivedCategory/Basic.lean:137]: Once the suitable instances are in place (for which we need this `Q (C := C)`, we shall mostly write `Q.obj K` with `K : CochainComplex C _`. Then, the `C` can be easily infered.

File: Mathlib/Algebra/Homology/DerivedCategory/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+/-- The derived category obtained using the constructed localized category of cochain complexes
+with respect to quasi-isomorphisms. This should be used only while proving statements
+which do not involve the derived category. -/
+def HasDerivedCategory.standard : HasDerivedCategory.{max u v} C :=
+  MorphismProperty.HasLocalization.standard _
+
+variable [HasDerivedCategory.{w} C]
+
+/-- The derived category of an abelian category. -/
+def DerivedCategory := HomologicalComplexUpToQuasiIso C (ComplexShape.up ℤ)
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/Algebra/Homology/DerivedCategory/Basic.lean]: Thanks. It is `Type (max u v)`, but I have made this more visible in the declaration, as well as for the `Category.{w}` instance.

**Review by jcommelin — APPROVED** [review]
Thanks :tada: This is a massive milestone! Huge kudos!

bors merge

**Pull request closed without merging**